### PR TITLE
[4.0] Add support to unpublish module from Control Panel

### PR DESCRIPTION
**Pull request by C-Lodder**
### Summary of Changes

This is a continuation of the initial prototype I did: https://github.com/joomla/40-backend-template/pull/251

In the control panel, we have a cog icon that directs you to the settings for that module. Now, it displays a dropdown with a new option to directly unpublish the module. Upon doing so, it will instantly disappear from the display.

### Testing Instructions

1. Fetch the code
2. Run `npm i` (apologies)
3. Login to the your Joomla admin
4. Click the "cog" icon for any module and select "Unpublish" from the dropdown

### Note

There are some changes to other files which keeps the dropdowns in other areas of the admin template intact.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_cpanel.ini [hunk 1]
```diff
@@ -24,5 +24,7 @@ COM_CPANEL_TITLE_SYSTEM_PANEL="System Panel"
...
 COM_CPANEL_MSG_UPDATEDEFAULTSETTINGS_TITLE="Updated site security recommendations"
 COM_CPANEL_WELCOME_BEGINNERS_MESSAGE="<p>Community resources are available for new users.</p><ul><li><a href=\"https://docs.joomla.org/Special:MyLanguage/Portal:Beginners\">Joomla! Beginners Guide</a></li><li><a href=\"https://forum.joomla.org/viewforum.php?f=706\">New to Joomla! Forum</a></li></ul>"
+COM_CPANEL_UNPUBLISH_MODULE_ERROR="Error unpublishing the module"
+COM_CPANEL_UNPUBLISH_MODULE_SUCCESS="Module successfully unpublished"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_cpanel.ini]: Please remove the redundant word "successfully"

File: administrator/language/en-GB/en-GB.ini [hunk 2]
```diff
@@ -150,6 +150,7 @@ JACTION_LOGIN_OFFLINE="Offline Access"
 JACTION_LOGIN_SITE="Site Login"
 JACTION_MANAGE="Access Administration Interface"
 JACTION_OPTIONS="Configure Options Only"
+ACTION_UNPUBLISH="Unpublish"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: This should be JACTION_UNPUBLISH

File: administrator/templates/atum/html/modules.php [hunk 3]
```diff
@@ -65,14 +67,19 @@ function modChrome_well($module, &$params, &$attribs)
...
-			echo '<div class="module-actions">';
-			echo '<a href="' . Route::_('index.php?option=com_modules&task=module.edit&id=' . (int) $module->id)
-				. '"><span class="fa fa-cog"><span class="sr-only">' . Text::_('JACTION_EDIT') . " " . $module->title . '</span></span></a>';
+			$dropdownPosition = Factory::getLanguage()->isRTL() ? 'left' : 'right';
+
+			echo '<div class="module-actions dropdown">';
+			echo '<a data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" id="dropdownMenuButton-' . $id . '"><span class="fa fa-cog"><span class="sr-only">' . Text::_('JACTION_EDIT') . ' ' . $module->title . '</span></span></a>';
+			echo '<div class="dropdown-menu dropdown-menu-' . $dropdownPosition . '" aria-labelledby="dropdownMenuButton-' . $id . '">';
+			echo '<a class="dropdown-item" href="' . Route::_('index.php?option=com_modules&task=module.edit&id=' . $id) . '">' . Text::_('JACTION_EDIT') . '</a>';
+			echo '<a class="dropdown-item unpublish-module" data-module-id="' . $id . '">' . Text::_('ACTION_UNPUBLISH') . '</a>';
```
Comment (brianteeman) [inline administrator/templates/atum/html/modules.php]: Please change to JACTION_UNPUBLISH

File: administrator/templates/atum/html/modules.php [hunk 4]
```diff
@@ -106,14 +115,19 @@ function modChrome_body($module, &$params, &$attribs)
...
```
Comment (brianteeman) [inline administrator/templates/atum/html/modules.php]: Please change to JACTION...

**Comment by C-Lodder** [comment]
@brianteeman All done. Thanks

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on d88d16132fd31c6d90e98172f47a4769ae74a525<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23934">issues.joomla.org/tracker/joomla-cms/23934</a>.</sub>

**Comment by infograf768** [comment]
Drone throws errors for javascript-cs

File: administrator/language/en-GB/en-GB.com_cpanel.ini [hunk 5]
```diff
@@ -24,5 +24,7 @@ COM_CPANEL_TITLE_SYSTEM_PANEL="System Panel"
...
 COM_CPANEL_MSG_UPDATEDEFAULTSETTINGS_TITLE="Updated site security recommendations"
 COM_CPANEL_WELCOME_BEGINNERS_MESSAGE="<p>Community resources are available for new users.</p><ul><li><a href=\"https://docs.joomla.org/Special:MyLanguage/Portal:Beginners\">Joomla! Beginners Guide</a></li><li><a href=\"https://forum.joomla.org/viewforum.php?f=706\">New to Joomla! Forum</a></li></ul>"
+COM_CPANEL_UNPUBLISH_MODULE_ERROR="Error unpublishing the module"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_cpanel.ini:27]: Please sort keys in alpha order.

Comment (C-Lodder) [reply administrator/language/en-GB/en-GB.com_cpanel.ini:27]: done

**Comment by coolcat-creations** [comment]
Thats a great improvement,
I tested it but the unpublish link does not do anything. The edit link works.

I'd open a seperate issue with following problem:
If you edit the module and close it, you end up in the module manager. It would be good to implement a refererlink on such icons so when you close something you get back to where you came from.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23934">issues.joomla.org/tracker/joomla-cms/23934</a>.</sub>

**Comment by C-Lodder** [comment]
Thanks Elisa. I'll add the refer link.
Shame it doesnt work for you though. Do you get any console errors?

**Comment by brianteeman** [comment]
@coolcat-creations are you sure you ran npm i after applying the patch. If you didnt then the dropdown appears but the functionality behind it is not present

**Comment by infograf768** [comment]
Unpublish works fine here.
@C-Lodder 
Shall we wait for the redirect depending on the origin of the edit to RTC this patch?

**Comment by C-Lodder** [comment]
If that feature is the correct approach then sure

**Comment by brianteeman** [comment]
Regarding the "return"

This issue is common in very many places. eg adding a note to a user

**Comment by infograf768** [comment]
I do think indeed it makes sense to redirect to CPanel when editing the module from there.
Now, tbh, this new feature implies a lot of code for not much gain. But that's only me.

**Comment by C-Lodder** [comment]
I'll leave it up to your guys then. RTC if you want or I can add the redirect in.

**Comment by infograf768** [comment]
+1 for redirecting to CPanel.
It is not because it does not work elsewhere that we should not do it when we think of it

**Comment by brianteeman** [comment]
> It is not because it does not work elsewhere that we should not do it when we think of it

Just pointing out that it needs to be done in multiple places not just here

**Comment by C-Lodder** [comment]
On second thoughts, adding a redirect is out of scope of this PR. This is purely to add support for unpublishing a module via the front-end and does not touch how the edit functionality works.

@coolcat-creations @infograf768 Please create a separate issue

**Comment by C-Lodder** [comment]
@wilsonge conflicts resolved

**Comment by C-Lodder** [comment]
@wilsonge conflicts resolved again

**Comment by infograf768** [comment]
Restarted drone...

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 952d5831ef20c693fd3e98838e9c617d95d41b22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23934">issues.joomla.org/tracker/joomla-cms/23934</a>.</sub>

**Comment by ghost** [comment]
@brianteeman can you please retest?

File: administrator/templates/atum/scss/vendor/bootstrap/_dropdown.scss [hunk 6]
```diff
@@ -9,6 +9,14 @@
   }
 }
 
+.dropdown-menu {
+  box-shadow: $atum-box-shadow;
+}
+
+.dropdown-item {
+	cursor: pointer;
```
Comment (Quy) [inline administrator/templates/atum/scss/vendor/bootstrap/_dropdown.scss]:
```suggestion
  cursor: pointer;
```

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on a1b150b2085a59f9e8af95c191af5fccfe4a3cf3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23934">issues.joomla.org/tracker/joomla-cms/23934</a>.</sub>

**Comment by infograf768** [comment]
@C-Lodder 
Please test so that we can merge this and go on with the return

**Comment by infograf768** [comment]
@coolcat-creations 
Please test so that we can go on with the return.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 52fb2cb5a2d16195ff8ccaad37b5d3d77557ca3c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23934">issues.joomla.org/tracker/joomla-cms/23934</a>.</sub>

**Comment by wilsonge** [comment]
Thankyou!

**Comment by infograf768** [comment]
Please test https://github.com/joomla/joomla-cms/issues/24109 for the redirection.

**Pull request merged**